### PR TITLE
Standardize error responses for failed authentication

### PR DESCRIPTION
**Pull request by gigxz**
When any of these endpoints are hit without authentication (ie cookie is missing or invalid, or CSRF token is missing or invalid), the server responds with a `401` with response body `{ error: { type: 'unauthenticated' } }` (at minimum).

This covers all the HMIS endpoints:
* /hmis/user
* /hmis/login
* /hmis/logout
* /hmis/hmis-gql

### CODE REVIEW

File: lib/devise/custom_auth_failure.rb [hunk 1]
```diff
@@ -6,7 +6,7 @@
 
 class CustomAuthFailure < Devise::FailureApp
```
Comment (gigxz) [inline lib/devise/custom_auth_failure.rb:7]: @ZachCMP @eanders as of now, this determines how authentication errors are returned from the GraphQL endpoint.

GraphQL is not REST, though, and Apollo has it's own spec for error handling: https://www.apollographql.com/docs/apollo-server/data/errors/#unauthenticated

If we were to follow that approach exactly, the authentication errors should look different for `/hmis/hmis-gql`. Something like a 200 with a JSON blob and 'Extension' code equal to `UNAUTHENTICATED`.

In other words.. we are treating authentication failure as a "server error" instead of a "graphql error" here. Maybe later on if we have an apollo graphql federation server, it will look different(?) but I think this approach makes sense for now.

File: drivers/hmis/app/controllers/hmis/sessions_controller.rb [hunk 2]
```diff
@@ -5,11 +5,16 @@
...
 class Hmis::SessionsController < Devise::SessionsController
+  include Hmis::Concerns::JsonErrors
   include AuthenticatesWithTwoFactor
   respond_to :json
   skip_before_action :verify_signed_out_user
   before_action :authenticate_with_2fa, only: [:create], if: -> { two_factor_enabled? }
 
+  rescue_from ActionController::InvalidAuthenticityToken do
+    render_json_error(401, :unauthenticated)
+  end
```
Comment (gigxz) [inline drivers/hmis/app/controllers/hmis/sessions_controller.rb]: Because we use `protect_from_forgery with: :exception`, devise raises an `ActionController::InvalidAuthenticityToken` exception when unable to validate the csrf token (happens if missing or invalid csrf token _or_ session cookie)

Right now, that leads to a 422 response (in development) or a 500 response (in prod - I think). We really want to return a 401 here, though, because the user is effectively not authenticated. That lets us have simpler logic on the frontend: if any request returns a 401, we need to redirect the user to log back in.

Comment (eanders) [reply drivers/hmis/app/controllers/hmis/sessions_controller.rb]: Maybe we can just `protect_from_forgery with: :some_method_name`  This could be done at the root HMIS level or root GraphQL level if that makes sense.
https://api.rubyonrails.org/classes/ActionController/RequestForgeryProtection/ClassMethods.html
`:reset_session` might give what you're looking for.

Comment (gigxz) [reply drivers/hmis/app/controllers/hmis/sessions_controller.rb]: Thanks @eanders. I think `reset_session` is what we want for the base controller (used for graphql), so I updated that. 
For the sessions controller we need something different, because `reset_session` would allow the user to log in even if the login form does not have a csrf token. I updated the sessions controller to use `handle_unverified_request` instead of `rescue_from`, which should have the same effect but looks cleaner.

Comment (eanders) [reply drivers/hmis/app/controllers/hmis/sessions_controller.rb]: Excellent!

**Review by eanders — APPROVED** [review]
Maybe we should write a few tests to confirm non-logged in users returns the expected json response?

**Comment by gigxz** [comment]
@eanders definitely. I added a couple tests to the sessions controller spec to test (1) if user tries to log in without csrf token, it fails, and (2) if an inactive user tries to login, the reason is included in the response. The second one works because of the changes to `CustomAuthFailure` in this PR.


I think we should also have tests for this flow:
1) succcessful login
2) user cookie is invalidated
3) post graphql request (should fail)

But it will require some test infra setup to get the graphql endpoint working. I think @ZachCMP was planning to add rspec tests for graphql, so I'll hold off there for now
